upup/pkg/fi/cloudup/scaleway/cloud.go

```diff
@@ -16,7 +16,395 @@ limitations under the License.
 
 package scaleway
 
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	account "github.com/scaleway/scaleway-sdk-go/api/account/v2alpha1"
+	"github.com/scaleway/scaleway-sdk-go/api/instance/v1"
+	"github.com/scaleway/scaleway-sdk-go/scw"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	kopsv "k8s.io/kops"
+	"k8s.io/kops/dnsprovider/pkg/dnsprovider"
+	"k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/pkg/cloudinstances"
+	"k8s.io/kops/upup/pkg/fi"
+)
+
 const (
 	TagClusterName      = "kops.k8s.io/cluster"
 	KopsUserAgentPrefix = "kubernetes-kops/"
+	TagInstanceGroup    = "instance-group"
+	TagNameRolePrefix   = "k8s.io/role/"
 )
+
+// ScwCloud exposes all the interfaces required to operate on Scaleway resources
+type ScwCloud interface {
+	fi.Cloud
+
+	ClusterName(tags []string) string
+	DNS() (dnsprovider.Interface, error)
+	ProviderID() kops.CloudProviderID
+	Region() string
+	Zone() string
+
+	AccountService() *account.API
+	InstanceService() *instance.API
+
+	DeleteGroup(group *cloudinstances.CloudInstanceGroup) error
+	DeleteInstance(i *cloudinstances.CloudInstance) error
+	DeregisterInstance(instance *cloudinstances.CloudInstance) error
+	DetachInstance(instance *cloudinstances.CloudInstance) error
+	FindClusterStatus(cluster *kops.Cluster) (*kops.ClusterStatus, error)
+	FindVPCInfo(id string) (*fi.VPCInfo, error)
+	GetApiIngressStatus(cluster *kops.Cluster) ([]fi.ApiIngressStatus, error)
+	GetCloudGroups(cluster *kops.Cluster, instancegroups []*kops.InstanceGroup, warnUnmatched bool, nodes []v1.Node) (map[string]*cloudinstances.CloudInstanceGroup, error)
+
+	GetClusterServers(clusterName string, serverName *string) ([]*instance.Server, error)
+	GetClusterVolumes(clusterName string) ([]*instance.Volume, error)
+
+	DeleteServer(server *instance.Server) error
+	DeleteVolume(volume *instance.Volume) error
+}
+
+// static compile time check to validate ScwCloud's fi.Cloud Interface.
+var _ fi.Cloud = &scwCloudImplementation{}
+
+// scwCloudImplementation holds the scw.Client object to interact with Scaleway resources.
+type scwCloudImplementation struct {
+	client *scw.Client
+	region scw.Region
+	zone   scw.Zone
+	tags   map[string]string
+
+	accountAPI  *account.API
+	instanceAPI *instance.API
+}
+
+// NewScwCloud returns a Cloud with a Scaleway Client using the env vars SCW_ACCESS_KEY, SCW_SECRET_KEY,
+// SCW_DEFAULT_PROJECT_ID, SCW_DEFAULT_REGION and SCW_DEFAULT_ZONE
+func NewScwCloud(tags map[string]string) (ScwCloud, error) {
```

Comment:
Maybe use `errors.NewAggregate(errs)` from k8s.io/apimachinery/pkg/util/errors to go through each of these checks in one go and return an aggregate error. Tends to be more user friendly than having the user error on one env, fix, then error on the next and repeat.